data_diff/dbt.py

```diff
@@ -0,0 +1,313 @@
+import os
+from typing import List
+import requests
+import json
+import yaml
+import rich
+import logging
+
+from dataclasses import dataclass
+from packaging.version import parse as parse_version
+from pathlib import Path
+from dbt_artifacts_parser.parser import parse_run_results, parse_manifest
+from . import connect_to_table, diff_tables, Algorithm
+from dbt.config.renderer import ProfileRenderer
+
+CONST_RUN_RESULTS_PATH = "/target/run_results.json"
+CONST_MANIFEST_PATH = "/target/manifest.json"
+CONST_PROJECT_FILE = "/dbt_project.yml"
+CONST_PROFILES_FILE = "/profiles.yml"
+CONST_DBT_V_ABOVE = "1.0.0"
+CONST_DBT_V_BELOW = "1.4.0"
+
+
+@dataclass
+class DiffVars:
+    dev_path: List[str]
+    prod_path: List[str]
+    primary_keys: List[str]
+    datasource_id: str
+    connection: dict
+
+
+class DbtDiffer:
+    @classmethod
+    def diff(self, profiles_dir_override: str = None, project_dir_override: str = None, is_cloud: bool = False):
+        dbt_parser = DbtParser(profiles_dir_override, project_dir_override, is_cloud)
+        models = dbt_parser.get_models()
+        dbt_parser.set_project_dict()
+        datadiff_variables = dbt_parser.get_datadiff_variables()
+        config_prod_database = datadiff_variables.get("prod_database")
+        config_prod_schema = datadiff_variables.get("prod_schema")
+        datasource_id = datadiff_variables.get("datasource_id")
+
+        if not is_cloud:
+            dbt_parser.set_connection()
+
+        assert (
+            config_prod_database is not None or config_prod_schema is not None
+        ), "Expected a value for prod_database: or prod_schema: under \nvars:\n  data_diff: "
```

Comment:
It is better to use explicit `if` & `raise` for input validation or business logic checks. Assertions can be disabled if the users prefer to run Python in "optimized" mode (`python -O`), which can be the case if the library is used not via CLI but is embedded into the users' toolchain (for example, I used it for data diffs in Alembic migrations, though it was not for dbt):

```
$ python -O -c 'assert False'
(no output)
```

```
$ python -c 'assert False'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
AssertionError
```

Reply:
[a0d7980](https://github.com/datafold/data-diff/pull/364/commits/a0d7980dec42969c50f339c4824ac3a0f02f08e9)